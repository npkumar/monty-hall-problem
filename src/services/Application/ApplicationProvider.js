import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

// Actions
const SET_SIMULATIONS = "SET_SIMULATIONS";
const SET_IS_CHANGE_DOOR = "SET_IS_CHANGE_DOOR";
const SET_WIN_PERCENTAGE = "SET_WIN_PERCENTAGE";
const SET_IS_SIMULATION_RUNNING = "SET_IS_SIMULATION_RUNNING";

export const ApplicationContext = createContext(null);

function reducer(state, action) {
  switch (action.type) {
    case SET_SIMULATIONS: {
      return { ...state, simulations: action.payload, winPercentage: null };
    }
    case SET_IS_CHANGE_DOOR: {
      return { ...state, isChangeDoor: action.payload, winPercentage: null };
    }
    case SET_WIN_PERCENTAGE: {
      return { ...state, winPercentage: action.payload };
    }
    case SET_IS_SIMULATION_RUNNING: {
      return { ...state, isSimulationRunning: action.payload };
    }
    default:
      throw new Error();
  }
}
const ApplicationProvider = ({ children }) => {
  const initialState = {
    simulations: 4000,
    isChangeDoor: true,
    winPercentage: null,
    isSimulationRunning: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setSimulations: (simulations) =>
      dispatch({ type: SET_SIMULATIONS, payload: simulations }),
    setIsChangeDoor: (isChangeDoor) =>
      dispatch({ type: SET_IS_CHANGE_DOOR, payload: isChangeDoor }),
    setWinPercentage: (winPercentage) =>
      dispatch({ type: SET_WIN_PERCENTAGE, payload: winPercentage }),
    setIsSimulationRunning: (isSimulationRunning) =>
      dispatch({
        type: SET_IS_SIMULATION_RUNNING,
        payload: isSimulationRunning,
      }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
