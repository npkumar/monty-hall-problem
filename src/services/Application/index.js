// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";
