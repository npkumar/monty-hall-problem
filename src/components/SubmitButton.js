import React from "react";
import { withStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import { green } from "@material-ui/core/colors";
import { useApplication } from "../services/Application";
import api from "../utils/api";

const ColorButton = withStyles((theme) => ({
  root: {
    color: theme.palette.getContrastText(green[500]),
    backgroundColor: "#52af77",
    "&:hover": {
      backgroundColor: green[700],
    },
  },
}))(Button);

const SubmitButton = () => {
  const {
    application: { isSimulationRunning, simulations, isChangeDoor },
    setWinPercentage,
    setIsSimulationRunning,
  } = useApplication();

  const handleSubmit = async () => {
    try {
      setIsSimulationRunning(true);
      const { data } = await api.post(
        "https://mhpnpk-server.herokuapp.com/simulate",
        {
          simulations,
        }
      );
      const { winPercentageWithoutSwitch, winPercentageAfterSwitch } = data;
      setWinPercentage(
        isChangeDoor ? winPercentageAfterSwitch : winPercentageWithoutSwitch
      );
      setIsSimulationRunning(false);
    } catch (err) {
      setIsSimulationRunning(false);
      setWinPercentage(null);
      console.log(err);
    }
  };

  return (
    <ColorButton
      fullWidth
      size="large"
      variant="contained"
      color="primary"
      onClick={handleSubmit}
      disabled={isSimulationRunning}
      data-testid="button"
    >
      Run Simulation
    </ColorButton>
  );
};

export default SubmitButton;
