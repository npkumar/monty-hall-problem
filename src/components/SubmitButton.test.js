import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import SubmitButton from "./SubmitButton";
import { ApplicationContext } from "../services/Application/ApplicationProvider";

const value = {
  application: {
    isSimulationRunning: false,
    simulations: 4000,
    isChangeDoor: true,
  },
};

describe("SubmitButton", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <SubmitButton />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have simulations and slider", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <SubmitButton />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("button")).toBeInTheDocument();
    expect(getByTestId("button")).toHaveTextContent("Run Simulation");
  });
});
