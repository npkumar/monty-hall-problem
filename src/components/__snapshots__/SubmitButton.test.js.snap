// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitButton renders 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-1 MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
    data-testid="button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Run Simulation
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
